File: src/visual-graph/builder/state-ref-id.ts
import type { BuilderContext } from "./context.js";
import { nodeId } from "./node-id.js";
import { stateAt } from "./state-at.js";
import { writeOpNodeId } from "./write-op-node-id.js";

export function stateRefId(
  refId: string,
  varId: string,
  ctx: BuilderContext,
): string {
  const op = ctx.writeOpByRef.get(refId);
  if (op) {
    return writeOpNodeId(op.refId);
  }
  const ref = ctx.ir.references.find((r) => r.id === refId);
  if (!ref) {
    return nodeId(varId);
  }
  const stateRef = stateAt(
    varId,
    ref.identifier.span.offset,
    ctx.writeOpsByVariable,
  );
  return stateRef === varId ? nodeId(varId) : writeOpNodeId(stateRef);
}

File: crates/unsnarl-visual-graph/src/builder/state_ref_id.rs
//! Mirrors `ts/src/visual-graph/builder/state-ref-id.ts`.
//!
//! Resolve the *state* of `varId` at the moment ref `refId` reads
//! it: a Write op's node id when one exists for the ref itself,
//! the most recent preceding write otherwise, or the variable
//! declaration node when no write has happened yet.

use super::context::BuilderContext;
use super::node_id::node_id;
use super::state_at::state_at;
use super::write_op_node_id::write_op_node_id;

pub fn state_ref_id(ref_id: &str, var_id: &str, ctx: &BuilderContext<'_>) -> String {
    if let Some(op) = ctx.write_op_by_ref.get(ref_id) {
        return write_op_node_id(&op.ref_id);
    }
    let Some(r) = ctx.ir.references.iter().find(|r| r.id.value() == ref_id) else {
        return node_id(var_id);
    };
    let state_ref = state_at(
        var_id,
        r.identifier.span().offset.0,
        &ctx.write_ops_by_variable,
    );
    if state_ref == var_id {
        node_id(var_id)
    } else {
        write_op_node_id(&state_ref)
    }
}
